### ISSUE TITLE
Questions on training EpiAgent with a 1.35M-token embedding + huge decoder

### ISSUE THREAD

**Issue by super-dainiu** [xy-chen16/EpiAgent#9]
Hi authors, 

Thanks for this great project. I’m trying to understand the practical systems recipe for handling the extremely large token space used here (vocab_size=1,355,449).

In the current implementation, there are two particularly heavy components:
	1.	Token embedding table
```
self.cCRE_embedding = nn.Embedding(vocab_size, embedding_dim)  # 1,355,449 x 512
```
This alone is ~1.355M * 512 ≈ 694M parameters.
	2.	Signal reconstruction head
```
self.signal_decoder = nn.Linear(embedding_dim, vocab_size - 4)  # 512 -> ~1.355M
```
That is another ~512 * 1.355M ≈ 694M parameters (+ bias).

So you end up with ~1.4B parameters before counting the Transformer, and the embedding/decoder dominate both memory and optimizer state.

Could you clarify what strategy you used in practice?

•	Are you using ZeRO-3 / FSDP with parameter + optimizer state sharding?
•	Do you tie signal_decoder.weight with cCRE_embedding.weight (common in LM heads)?
•	Is vocab_size=1,355,449 the number of cCREs? If so:
	•	Do you prune rare cCREs or standardize cCREs with original fragments?

**Comment by xy-chen16** [comment]
Hi,

Sorry for the delayed response — I’ve been in the middle of job applications recently, which slowed down my replies a bit.

Regarding your questions: at this stage, we did not apply many additional engineering-level optimizations beyond what is shown in the released code. The publicly available implementation largely reflects the parameter settings and model structure we used in practice.

As for the vocabulary size: vocab_size = 1,355,449 does not exactly equal the number of cCREs. The actual number of cCREs is 1,355,445, and the remaining four tokens correspond to special tokens (e.g., padding / masking–related tokens), which is also described in the paper.

Thank you again for the careful reading and for raising these thoughtful questions.

Best regards,
Xiaoyang

**Comment by super-dainiu** [comment]
Thank you very much for your detailed reply.

I have a quick follow-up question regarding vocabulary size. Using around 1.35 million cCRE tokens is quite uncommon in modeling and system design, even for large-scale language models. I was curious if you considered or tested methods to reduce the vocabulary during your development process.

For example:
- Removing very rare or low-information cCREs,
- Merging highly correlated regions,
- Or applying dimensionality reduction techniques (such as LSI) to ATAC features.

I understand from the paper that intentionally not filtering cCREs was a deliberate decision and that modeling the full regulatory space is a core goal of EpiAgent. However, I wonder if you observed or deliberately chose not to explore trade-offs related to vocabulary size, such as its effects on training loss, convergence, or downstream outcomes.

Currently, I am pretraining a BERT-style model on ATAC-seq data, exploring how vocabulary pruning affects optimization and loss. I tested reducing the ATAC feature space using an LSI-style projection, removing ~3/4 of the cCRE dimensions. Surprisingly, despite a smaller effective vocabulary, the model converges more stably and reaches lower training loss than the baseline with the full cCRE vocabulary. A training loss comparison figure illustrates this.

<img width="2372" height="1771" alt="Image" src="https://github.com/user-attachments/assets/dd6e4c40-93c7-417d-b566-42da7f3459ab" />

I’d greatly appreciate your perspective on whether such reductions would be reasonable or conflict with the goals of EpiAgent.

Best,
Daniel

**Comment by xy-chen16** [comment]
Hi Daniel,

Thank you for the thoughtful follow-up and for sharing your experimental observations — the results you describe are very interesting.

We agree that the strategies you mentioned (e.g., pruning rare cCREs, merging correlated regions, or applying LSI-style dimensionality reduction) are all technically valid and often effective from an optimization and systems perspective.  As your results suggest, such approaches can indeed improve training stability and reduce loss by simplifying the feature space.

In our case, the decision to retain the full cCRE vocabulary was largely intentional and conceptually driven.  We aimed to model the chromatin regulatory landscape at its original resolution, treating each cCRE as a first-class symbolic unit.  From this perspective, vocabulary reduction can be viewed as an engineering-oriented optimization that trades off part of the biological signal — or what one might call the “biological granularity / interpretability” — for computational efficiency and smoother optimization.

So we see this primarily as a trade-off rather than a strictly right-or-wrong choice:
•	Vocabulary reduction may yield better convergence behavior and lower training loss.
•	Retaining the full cCRE space preserves maximal regulatory resolution and biological expressiveness, which was a core design goal of EpiAgent.


Thanks again for sharing your insights and results — it’s great to see parallel efforts probing these questions from different angles.

Best regards,
Xiaoyang
